docs/knowledgebase/advanced/try-runtime.md

```diff
@@ -0,0 +1,190 @@
+    ```
+
+To call into these hooks with `try-runtime`, a cargo `feature` is used. This requires specifying a `features` 
+dependency in your projects' relevant `Cargo.toml` files. With this cargo feature flag enabled, `try-runtime` can call into `OnRuntimeUpgrade` hooks using the 
+`#[cfg(feature = "try-runtime")]` macro which gives it access to the `OnRuntimeUpgrade` methods.
+
+> Refer to this how-to guide on integrating `try-runtime` to your project. _(TODO)_
+
+For example, in a storage migration, `try-runtime` can allow developers to 
+examine state both before and after a runtime upgrade:
+
+```rust
+    #[cfg(feature = "try-runtime")]
+    fn pre_upgrade() -> Result<(), &'static str> { Ok(()) }
+
+    #[cfg(feature = "try-runtime")]
+    fn post_upgrade() -> Result<(), &'static str> { Ok(()) }
+```
+
+> See how the Executive pallet implements the `try_runtime_upgrade` method [here][executive-example-frame] to 
+> aggregate all `OnRuntimeUpgrade` calls from a runtime and return a `Weight`, which could be used for testing.  
+
+### Helper functions
+
+[`OnRuntimeUpgradeHelpersExt`][oru-helpers-ext-rustdocs] are a set of helper functions made available from 
+[`frame_support::hooks`][hooks-rustdocs] in order to use `try-runtime` for testing storage migrations. These include:
+
+- **`storage_key`**: Generates a storage key unique to this runtime upgrade. This can be used to communicate data from pre-upgrade to post-upgrade state and check them.
+- **`set_temp_storage`**: Writes some temporary data to a specific storage that can be read (potentially in the post-upgrade hook).
+- **`get_temp_storage`** : Gets temporary storage data written by `set_temp_storage`.
+
+Using the [`frame_executive::Executive`][executive-rustdocs] struct, these helper functions in action would look like:
+
+```rust
+pub struct CheckTotalIssuance;
+impl OnRuntimeUpgrade for CheckTotalIssuance {
+    #[cfg(feature = "try-runtime")]
+    fn post_upgrade() {
+        // iterate over all accounts, sum their balance and ensure that sum is correct.
+    }
+}
+
+pub struct EnsureAccountsWontDie;
+impl OnRuntimeUpgrade for EnsureAccountsWontDie {
+    #[cfg(feature = "try-runtime")]
+    fn pre_upgrade() {
+        let account_count = frame_system::Accounts::<Runtime>::iter().count();
+        Self::set_temp_storage(account_count, "account_count");
+    }
+
+    #[cfg(feature = "try-runtime")]
+    fn post_upgrade() {
+        // ensure that this migration doesn't kill any account.
+        let post_migration = frame_system::Accounts::<Runtime>::iter().count();
+        let pre_migration = Self::get_temp_storage::<u32>("account_count");
+        ensure!(post_migration == pre_migration, "error ...");
+    }
+}
+
+pub type CheckerMigrations = (EnsureAccountsWontDie, CheckTotalIssuance);
+pub type Executive = Executive<_, _, _, _, (CheckerMigrations)>;
+```
+
+### CLI interface
+
+To use `try-runtime` from the command line, run your node with the `--features=try-runtime` flag. 
```

Comment:
It might be worth mentioning that `cargo run` is used here or specifying that the `try-runtime` subcommand is used when invoking the binary directly.

Reply:
> > Thanks for putting this together.
> > I'm interested in using `try-runtime` to ensure that our upcoming migrations go as expected. While there is a lot that can be done with the additional `pre_upgrade()` and `post_upgrade()` hooks, I would like to use `try-runtime` to begin > > Is this possible with `try-runtime`? If so, I'd love to see it mentioned in documentation.
> 
> Ah, it was this that I was thinking of: https://github.com/maxsam4/fork-off-substrate
> 
> I'm yet to play with it, but it looks incredibly useful. It might be worth mentioning in `try-runtime`'s documentation.

Hi Stephen - thanks for your contribution and comments here, I'll make an attempt to make this a little clearer in the article here. The answer is that that `try-runtime` seems like a useful tool in the scenario you're describing. Unlike 'fork-off-substrate', you can write unit tests in the runtime of your chain or specific pallets in `on_runtime_upgrade`.